### PR TITLE
splunk: add splunk-hec-raw() and splunk-hec-event() drivers

### PR DESCRIPTION
**Pull request by alltilla**
splunk: Added support for sending messages to Splunk HEC.

The `splunk-hec-event()` destination feeds Splunk via the [HEC events API](https://docs.splunk.com/Documentation/Splunk/9.0.4/RESTREF/RESTinput#services.2Fcollector.2Fevent.2F1.0).

Minimal config:
```
destination d_splunk_hec_event {
  splunk-hec-event(
    url("https://localhost:8088")
    token("70b6ae71-76b3-4c38-9597-0c5b37ad9630")
  );
};
```

Additional options include:
  * `event()`
  * `index()`
  * `source()`
  * `sourcetype()`
  * `host()`
  * `time()`
  * `default-index()`
  * `default-source()`
  * `default-sourcetype()`
  * `fields()`
  * `extra-headers()`
  * `extra-queries()`
  * `content-type()`

`event()` accepts a template, which declares the content of the log message sent to Splunk.
By default we send the `${MSG}` value.

`index()`, `source()`, `host()` and `time()` accepts templates, which declare the respective field
of each log message based on the set template.

`default-index()`, `default-source()` and `default-sourcetype()` accepts literal strings, which
will be used for fallback values if a log message does not set these fields. These values are
passed to the URL as query parameters, so they do not inflate the body of the HTTP request
for each message in the batch, which saves bandwidth.

`fields()` accepts template, which is passed as additional indexing metadata to Splunk.

`extra-headers()`, `extra-queries()` and `content-type()` are additional HTTP request options.

The `splunk-hec-raw()` destination feeds Splunk via the [HEC raw API](https://docs.splunk.com/Documentation/Splunk/9.0.4/RESTREF/RESTinput#services.2Fcollector.2Fraw.2F1.0).

Regarding its options, it is similar to the `splunk-hec-event()` destination with the addition
of the mandatory `channel()` option, which accepts a GUID as a literal string that differentiates
data from different clients. Another difference is that instead of `event()`, here the `template()`
option sets the template which will represent the log message in Splunk.

Minimal config:
```
destination d_splunk_hec_raw {
  splunk-hec-raw(
    url("https://localhost:8088")
    token("70b6ae71-76b3-4c38-9597-0c5b37ad9630")
    channel("05ed4617-f186-4ccd-b4e7-08847094c8fd")
  );
};
```

Signed-off-by: Attila Szakacs <attila.szakacs@axoflow.com>

### CODE REVIEW

**Comment by alltilla** [comment]
Tested manually on a local `splunk:latest` docker container with `peer-verify(no)`, but it would be nice of someone else could test it.

File: scl/splunk/splunk.conf [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  sourcetype("")
+  host("${HOST}")
+  time("${S_UNIXTIME}.${S_MSEC}")
+  fields("")
+
+  event("${MSG}")
+
+  batch_lines(5000)
+  batch_bytes(4096kB)
+  batch_timeout(300)
```
Comment (MrAnno) [inline scl/splunk/splunk.conf:89]: Splunk had a very small `max_content_length` limit in older versions.
It's now 800MB.

It's probably not worth lowering these defaults for old versions.

File: scl/splunk/splunk.conf [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  timeout(10)
+
```
Comment (MrAnno) [inline scl/splunk/splunk.conf:91]: Can we add `workers()` with a non-1 default?

File: scl/splunk/splunk.conf [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+              source="`source`"
+              sourcetype="`sourcetype`"
+              host="`host`"
+              time="`time`"
+              event="`event`"
+              `fields`)'
+    )
+    batch-lines(`batch_lines`)
+    batch-bytes(`batch_bytes`)
+    batch-timeout(`batch_timeout`)
```
[the remark this answers is not in the document]
Comment (alltilla) [reply scl/splunk/splunk.conf:120]: Example 2 here shows nothing between the JSON objects:
https://docs.splunk.com/Documentation/Splunk/9.0.4/Data/HECExamples#Example_2:_Send_multiple_events_to_HEC

```
curl "https://mysplunkserver.example.com:8088/services/collector" \
    -H "Authorization: Splunk CF179AE4-3C99-45F5-A7CC-3284AA91CF67" \
    -d '{"event": "Pony 1 has left the barn"}{"event": "Pony 2 has left the barn"}{"event": "Pony 3 has left the barn", "nested": {"key1": "value1"}}'
```

File: scl/splunk/splunk.conf [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+};
+
+# https://docs.splunk.com/Documentation/Splunk/9.0.4/RESTREF/RESTinput#services.2Fcollector.2Fevent.2F1.0
+block destination splunk_hec_event(
+  url()
+  token()
+
+  default_index("main")
+  default_source("syslog-ng")
+  default_sourcetype("syslog")
```
[the remark this answers is not in the document]
Comment (alltilla) [reply scl/splunk/splunk.conf]: Thanks, fixed!

File: scl/splunk/splunk.conf [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    `__VARARGS__`
+  );
+};
+
+# https://docs.splunk.com/Documentation/Splunk/9.0.4/RESTREF/RESTinput#services.2Fcollector.2Fevent.2F1.0
+block destination splunk_hec_event(
+  url()
+  token()
+
+  default_index("main")
```
[the remark this answers is not in the document]
Comment (alltilla) [reply scl/splunk/splunk.conf:76]: Thanks! I have kept it `main`.

File: scl/splunk/splunk.conf [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+  http(
+    url("`url`/services/collector/event/1.0?index=`default_index`&source=`default_source`&sourcetype=`default_sourcetype``extra_queries`")
+    headers(
+      "Authorization: Splunk `token`"
+      "Content-Type: `content_type`"
+      "Connection: keep-alive"
+      `extra_headers`
+    )
+    body('$(format-json --scope none --omit-empty-values
```
[the remark this answers is not in the document]
Comment (alltilla) [reply scl/splunk/splunk.conf]: I have fixed this as we talked about it IRL. Thanks!

File: scl/splunk/splunk.conf (hunk 2 above)
Comment (alltilla) [reply scl/splunk/splunk.conf:94]: Yes, fixed :)

**Review by MrAnno — APPROVED** [review]

**Pull request merged**